src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs

```diff
@@ -296,5 +284,164 @@ internal SafeSslHandle(IntPtr validSslPointer, bool ownsHandle) : base(IntPtr.Ze
+                GC.SuppressFinalize(this);
+            }
+
+            ~ReadBioBuffer()
+            {
+                Dispose(false);
+            }
+        }
+
+        internal class WriteBioBuffer : IDisposable
+        {
+            private SafeBioHandle _bioHandle;
+            private GCHandle _handle;
+            private byte[] _byteArray;
+            private int _bytesWritten;
+            private bool _isHandshake;
+
+            internal WriteBioBuffer(SafeBioHandle bioHandle)
+            {
+                _bioHandle = bioHandle;
+                _handle = GCHandle.Alloc(this, GCHandleType.Normal);
+                Interop.Crypto.ManagedSslBio.BioSetGCHandle(_bioHandle, _handle);
+            }
+
+            public int BytesWritten => _bytesWritten;
+
+            public void SetData(byte[] buffer, bool isHandshake)
+            {
+                Debug.Assert(_byteArray == null);
+
+                _byteArray = buffer;
+                _bytesWritten = 0;
+                _isHandshake = isHandshake;
+            }
+
+            public int TakeBytes(out byte[] output)
+            {
+                output = _byteArray;
+                var bytes = _bytesWritten;
+                Reset();
+                return bytes;
+            }
+
+            public void Reset()
+            {
+                _bytesWritten = 0;
+                _byteArray = null;
+            }
+
+            public int Write(Span<byte> input)
+            {
+                // Only for the handshake do we dynamically allocate
+                // buffers for normal encrypt operations we use a fixed
+                // size buffer handed to us and loop to do all the needed
+                // writes. This should be changed for the handshake as well
+                // but will require more securechannel/sslstatus changes
+                if (_isHandshake)
+                {
+                    if (_byteArray == null)
+                    {
+                        _byteArray = new byte[input.Length];
+                        _bytesWritten = 0;
+                    }
+
+                    else if (_byteArray.Length - _bytesWritten < input.Length)
+                    {
+                        var oldSpan = new Span<byte>(_byteArray);
+                        _byteArray = new byte[input.Length + _bytesWritten];
+                        oldSpan.CopyTo(_byteArray);
+                    }
+                }
+
+                int bytesToWrite = Math.Min(input.Length, _byteArray.Length - _bytesWritten);
+                if (bytesToWrite < 1)
+                {
+                    // We need to return -1 to indicate that it is an async method and
+                    // and the write should retry later rather and a zero indicating EOF
+                    Interop.Crypto.BioSetWriteFlag(_bioHandle);
+                    return -1;
+                }
+
+                input.Slice(0, bytesToWrite).CopyTo(new Span<byte>(_byteArray, _bytesWritten));
```

Comment:
Unrelated to this PR, but since I'm thinking of it on this line, @terrajobst, @KrzysztofCwalina, this may just be me, but FYI I've gotten confused multiple times recently by Span's ctors.  There are several 2-arg constructors, e.g. (void*, int), (ref T, int), and (T[], int): for the first two, the int is a length, but for the last, it's an offset.  I've made a few mistakes based on confusing that int as a length for the T[] version, and I expect we'll see others get confused as well.